packages/astro/src/config/types.ts

```diff
@@ -0,0 +1,30 @@
+import type fsMod from 'node:fs';
+import type { AstroSettings, InjectedDts } from '../@types/astro.js';
+import { AstroError, AstroErrorData } from '../core/errors/index.js';
+
+// TODO: only keep astro.d.ts in Astro 5
+export const CODEGENDIR_BASE_DTS_FILE = 'types.d.ts';
+const RESERVED_FILE_NAMES = ['astro.d.ts', CODEGENDIR_BASE_DTS_FILE];
+
+export function injectDts({
+	codegenDir,
+	filename,
+	content,
+	bypassValidation = false,
+	fs,
+}: Pick<AstroSettings, 'codegenDir'> &
+	InjectedDts & { bypassValidation?: boolean; fs: typeof fsMod }) {
+	if (
+		!bypassValidation &&
+		(!filename.endsWith('.d.ts') || RESERVED_FILE_NAMES.includes(filename))
+	) {
+		throw new AstroError(AstroErrorData.InvalidInjectTypesFilename);
+	}
+
+	ensureCodegenDirExists({ codegenDir, fs });
+	fs.writeFileSync(new URL(filename, codegenDir), content);
+}
+
+export function ensureCodegenDirExists({ codegenDir, fs }: { codegenDir: URL; fs: typeof fsMod }) {
+	fs.mkdirSync(codegenDir, { recursive: true });
+}
```

Comment:
This function should do more things:
- it should handle errors. Sometimes, permissions issues might occur during FS operations, and we have to handle them. We should add a `try/catch`. Or, we should handle the error in the places where we call this function.
- it should **check** that the folders don't exist already. I believe Node.js errors if we try to create something that already exists

Reply:
Agreed concerning your first point. Regarding 2, see this review by Fry and let me know https://github.com/withastro/astro/pull/9952#discussion_r1501255184

Reply:
[For reference](https://nodejs.org/api/fs.html#fspromisesmkdirpath-options):

> Calling `fsPromises.mkdir()` when `path` is a directory that exists results in a rejection only when `recursive` is false.

The same applies for the callback-based API and the sync API